commitizen/commands/changelog.py

```diff
@@ -164,21 +164,17 @@ def __call__(self):
         # Don't continue if no `file_name` specified.
         assert self.file_name
 
-        tags = changelog.get_version_tags(self.scheme, git.get_tags()) or []
-
+        tags = (
+            changelog.get_version_tags(self.scheme, git.get_tags(), self.tag_format)
+            or []
+        )
         end_rev = ""
         if self.incremental:
             changelog_meta = self.changelog_format.get_metadata(self.file_name)
             if changelog_meta.latest_version:
-                latest_tag_version: str = bump.normalize_tag(
-                    changelog_meta.latest_version,
-                    tag_format=self.tag_format,
-                    scheme=self.scheme,
-                )
                 start_rev = self._find_incremental_rev(
-                    strip_local_version(latest_tag_version), tags
+                    strip_local_version(changelog_meta.latest_version), tags
                 )
-
```

Comment:
Hello,
Removing this logic is breaking the search for custom tags (with a tag-format like `example-${version}`).
The first bump (with `--changelog` option) will have no problem and it will create the `CHANGELOG.md` file but any bump afterward will result in an error: `commitizen.exceptions.NoRevisionError: No tag found to do an incremental changelog`.

This is due to the `_find_incremental_rev` function that look at similarity between a given tag and the tags list. But without the normalization, it calculate similarity between a stripped tag (without any format so like `0.1.1`) and the formatted tag (like `example-0.1.0`) resulting in a similarity below threshold.

So, except if you have another edge case that was justifying this removal, I think it should be kept.

Reply:
Hi,

Thanks for the feedback, I'll put this to a draft status and try to find a solution that doesn't break the current way of working. My gut feeling is it might need an additional flag maybe something like `--strict-tag-matching` what are your thoughts on taking that approach?

I guess the fact you found that error might mean there is a missing test case, I will try adding that first to replicate the error with my changes so I have something to validate against.

Reply:
I used your version with a revert on the selected part ([here](https://github.com/Lowaiz/commitizen/)), and it seems to works pretty damn well for me, that's why I was asking for the "why" of the change.
I was able to bump and generate for multiple cz.toml file and multiple tag-format (in a monorepo with 3 sub-packages).

The normalizing part was introduced by that [PR](https://github.com/commitizen-tools/commitizen/pull/566/files), exactly fixing the error I got. 

I really think that you just need to keep the normalizing part, and we should be OK.

Reply:
Hi,

Thanks for clarifying I misunderstood your point.

Reply:
Oh I see the problem.

The difference in behavior between `example-${version}` and  `${version}-example` is due to the regex used in the `changelog_format/markdown.py`.
On my example, the regex only match the version number, but with your it parse the whole tag, probably because the suffix `-example` match some condition (like `-alpha` match for a alpha release) in the regex.

My suggestion would be to use the tag_format to build a regex to extract the tag from the line, and then from the extracted tag, use the version regex to extract the version

Reply:
A few questions/points after thinking further on this before trying to make further changes. 

Tests of the changelog use examples that don't appear to match the format of those written by the commitizen, is this to cover that consumers started using commitizen with an existing changelog?

If I am understanding the logic correctly we pull the version out of the changelog by matching for the tag format in the changelog, then take that version to match against the tags to find the latest tag in the changelog to get the starting revision, again is this to handle changelings that might not have originally been created by commitizen? If not should it be safe to assume that we could just match on tag in the changelog to get the starting revision when doing incremental changelogs?

Reply:
Just to clarify why I'm on this PR, I'm not a maintainer of this project but I'm greatly interested by the monorepo feature, so I try to make it happen a bit faster. 

> Tests of the changelog use examples that don't appear to match the format of those written by the commitizen, is this to cover that consumers started using commitizen with an existing changelog?

I think yes, I saw a PR or Issue saying that, can't remember where.


> If I am understanding the logic correctly we pull the version out of the changelog by matching for the tag format in the changelog, then take that version to match against the tags to find the latest tag in the changelog to get the starting revision, again is this to handle changelings that might not have originally been created by commitizen? If not should it be safe to assume that we could just match on tag in the changelog to get the starting revision when doing incremental changelogs?

I can't really tell, for both questions. 
For the logic I think you're right.
But I think I see a flaw in the current logic:
- if the `tag_format` does not contain `$version` or `${version}`, the substitution does not include the regex group `version` so the extraction fails in the changelog metadat fetch.
- Another [ongoing PR](https://github.com/commitizen-tools/commitizen/pull/692) seems to start addressing this flow, but I think it does not cover all uses cases.
- if a `tag_format` does not contains the complete version infos (major, minor, patch, pre-rel, dev-rel), we can't make the back-matching from the tag to the version, ultimately, I really think the complete version should be used to generate changelog and not the formatted tag, or we need to ensure that the format contains all infos, and make a function to build a regex from the tag (a bit like the mentioned PR). Other possibility could be to not base the changelog generation on the available info inside the current changelog. (at least when bumping version, as we can get the lat git tag and the upcoming one, there is no need to base the last version to the version available in the changelog)

I think it's a complicated question for now, maybe the maintainers could help us a bit to lead to a future-proof solution: @woile @Lee-W @noirbizarre, reading the conversation, do you have opinion or an idea on what to do to better handle tag/tag_format and changelogs generation ?

Reply:
I'm still trying to read through all the discussion, but here're some short answers to the questions

> if the tag_format does not contain $version or ${version}, the substitution does not include the regex group version so the extraction fails in the changelog metadata fetch.

We should catch the case that `tag_format` does not contain a version.

> if a tag_format does not contains the complete version infos (major, minor, patch, pre-rel, dev-rel), we can't make the back-matching from the tag to the version, ultimately, I really think the complete version should be used to generate changelog and not the formatted tag, or we need to ensure that the format contains all infos, and make a function to build a regex from the tag (a bit like the mentioned PR). Other possibility could be to not base the changelog generation on the available info inside the current changelog. (at least when bumping version, as we can get the last git tag and the upcoming one, there is no need to base the last version to the version available in the changelog)

Yes, I think `we need to ensure that the format contains all infos, and make a function to build a regex from the tag ` is what we should do.

Reply:
Not sure whether this would be a breaking change 🤔 if that's the case, `--strict-tag-matching`  is something we could try.

Reply:
We might not need to do a breaking change for this (the functionality was broken already).

I have added handling for different tag_formats other than the standard `$version` to the markdown format. If someone could take a look and let me know if this is a valid approach I will implement the same for all the format types.

I have also taken a stab at documenting how to work with commitizen in a monorepo.

Reply:
Is there anything that needs clarification for this one? (I guess not?) or should I start reviewing?

Reply:
I just want to check the way I have handled the different possible tag_formats in markdown.py seems OK. If it is I will make the appropriate changes to the other formats taking the same approach. Once I have completed that it will be ready for review

Reply:
I don't see any major flaw at the first glance but will try to take a deeper look this weekend.

Reply:
Sorry it took a while for me to get this finished off. Ready for review when you have time.

Reply:
Thanks! I was not able to review it last week. 😞  Let's see whether I can at least check a portion of it this week. 💪